### ISSUE TITLE
Discovering Build Task from module, and import them

### ISSUE THREAD

**Issue by gaelcolas** [nightroman/Invoke-Build#47]
I really like Invoke-Build and the way you can split tasks in different files, use parameters and the `property` keyword to propagate them across dot sourced Build tasks files, providing defaults or overriding values.

I think some PS Modules should _export_ tasks generic but relevant to their subject (say [PSDepend](https://github.com/RamblingCookieMonster/PSDepend) could expose some tasks like those [here](https://github.com/gaelcolas/DscConfiguration/blob/master/.build/PSDepend/ResolveDependencies.PSDepend.build.ps1)), and then a user can import and compose them in a single `.build.ps1` file, overriding variables as needed.

This would allow much more re-usability from Build scripts, instead of everyone 'making their own'.

Then it would be good to have Invoke-Build providing some helper functions for discovering those Build tasks/file from other installed modules or from the gallery.

To implement that function, I think [Plaster](https://github.com/PowerShell/Plaster) has a very neat way of doing exactly that for PlasterTemplates.

In short, the module needs to 'expose' an [extension in its metadata](https://github.com/PowerShell/Plaster/blob/1e62113849f658b905c5b256537ecacd8ad1f967/examples/TemplateModule/README.md), then the Plaster module can [find the extension](https://github.com/PowerShell/Plaster/blob/master/src/GetModuleExtension.ps1) and [list its Templates](https://github.com/PowerShell/Plaster/blob/master/src/GetPlasterTemplate.ps1).

What do you think?

/cc @RamblingCookieMonster

**Comment by gaelcolas** [comment]
By the way, those who do not want to 'pollute' their module with build tasks can similarly create another module i.e. `TheirModuleName.BuildTasks` that depends on `TheirModuleName` and implements the extension as described above.
Might be even better... but raises the same question as Plaster and its templates [potentially polluting the PSGallery](https://github.com/PowerShell/Plaster/issues/224) (see Jaykul's comment)

**Comment by nightroman** [comment]
Thank you for the interesting topic. It is the area for some development and potential improvements. I have been thinking of something similar a few times. I did not find an elegant and universal way of doing this. So far.

I am looking for ideas, suggestions, and especially use cases. I am not keen to rush, though, let's discuss and think.

"Modules" (or other "things") may expose "special" tools in many ways. They do not have to wrap them in tasks. Pre-wrapping may be handy in some cases, yes. But then a user must use Invoke-Build. On the other hand: (a) not task-wrapped tools could be used in more PowerShell scenarios; (b) it is not that difficult for an end user to compose a task that just calls an exposed tool with proper parameters. The latter, in fact, is the primary/typical scenario of Invoke-Build, i.e. tools and complex code are outside, tasks just glue things together and declare dependencies.

**Comment by nightroman** [comment]
@gaelcolas 
A side question. In your example I noticed constructs

    $Parameter = $(try { property Name} catch { }), #propagate or use $null

Do you think it makes sense to allow the default null value if it is provided as null? The used construct is not neat, e.g. it may generate errors polluting the `$Error`.

**Comment by RamblingCookieMonster** [comment]
I'm not sure if it's the right place, but if there are a set of more general-purpose tasks that might be worth distributing via the gallery, I'd be happy to include tasks with the [BuildHelpers](https://github.com/RamblingCookieMonster/BuildHelpers) module - the module is already out on the gallery, so wouldn't pollute the gallery in that sense, and the idea for that module was re-usable tools for builds, so it fits the goal, if it makes sense and a sensible implementation could be designed.

As for per-module, very interesting idea - but yeah, not sure what the best vehicle would be if you went that route

Cheers!

**Comment by gaelcolas** [comment]
re @nightroman: I think the best option could be to allow setting the default to $null such as `$parameter = (property propName $null)` so that backward compatibility works.

What I did was a quick hack for my PoC while learning the tool.
I haven't looked at the source code yet.
However, the best way to throw if no value is set would be by adding the `[Parameter(Mandatory)]` attribute to the parameter.

re @nightroman and @RamblingCookieMonster, I think there is value in offering generic and re-usable tasks, along with some modules (BuildHelpers being an obvious one), but many CI/CD related modules would also make sense. It would allow a bit more standardization and community-driven improvement, instead of everyone building things their own way, and would also encourage beginners to start composing simple build 'workflow' with CI/CD by re-using basic tasks, and then extending it with other integration (i.e. uploading test results to AppVeyor).

The build Tasks would only be an 'Extra', and by no mean would replace a module... I truly recognize the benefit of modules, and providing the build tasks would merely provide some bricks and example of composition, that the user is free to compose as they wish in their own build script.

As for the scenarios, I found that most build scripts revolve around similar principles, and many excel in different parts.
Re-using someone else's script is nearly impossible as there's always some specificity not relevant to your case, merging two monoliths while removing some bits is a nightmare as they don't respect the same naming conventions and variable names...

Starting with smaller bricks, doing atomic tasks, and splitting them in separate files (in essence, the PowerShell _best practices_ for modules), would improve readability, re-usability and flexibility.
I found that it seems relevant to group them based on category, and usually the module they leverage. As an example, here are common tasks that would make sense:
- PlatyPS: Create the documentation based on .md files
- PSScriptAnalyzer: Static code Checking...
- BuildHelpers: manage many common tasks
- PSSlack: Built-in Notification on Build failure (no more dependent on the CI tool's integration)
- AppVeyor/Bamboo/Jenkins: Upload test results
- DscBuildHelpers: Similar to BuildHelpers but for DSC...

And so on...
I hope it's clear enough to show you the point, now I'm gonna try to write some of those tasks and compose them to see how it feels.
This is obviously only food for thoughts for the moment, but I'm keen on getting more brain on the matter. I know that some of the trouble we'll get will be with the standards used (build variable Names, environment setup...) but with the help of Plaster and the BuildHelpers, PSdepend, PSDeploy, and other we have an opportunity to build up instead of _from scratch_ every time...

**Comment by RamblingCookieMonster** [comment]
Yeah, probably would be tough to come up with something that everyone would be happy with and that would fit everyone's needs.

That said, I like the idea, and given that we control some of those repos, I'd be happy to help hash out a basic convention and implement it on a few - e.g. PSSlack might be fun.  At the very least, we could illustrate this in a consumable way (friendly blog post) that folks like myself (far from the actual-dev-world) might be more comfortable with.

Roman - we can take discussion elsewhere if you'd rather not have it in this repo : )

**Comment by nightroman** [comment]
Any ideas, suggestions, or new input on this? If this stalls I would rather close (we can open a new issue any time). Invoke-Build supports and encourages imported tasks, this is not a question. As far as I understand, the question is about some conventions on how modules/tools expose their shared tasks, so that Invoke-Build may find them. I do not have ideas or use cases on my own.

**Comment by gaelcolas** [comment]
Sorry, I was away the last couple of weeks.
I'll close this for now and will re-open when I have a 'demo' of what I have in mind, it's in progress but I'll be away again in 2 weeks so it might stall a bit until I'm back.

**Comment by doesitscript** [comment]
All of these are great points, I just wanted to add a few things.   

1. Having to have to dot source the build tasks in the default project build script is cumbersome when sharing build scripts and tasks with others. 
Suggest having a folder name and file naming convention that Invoke-Build includes in addition to the default build script when parsing tasks.

2. Variables are a limiting factor when sharing scripts.  
Suggest having a way to declare/describe variables used within a task script. Having a require statement depending on something like BuildHelpers would be a good start. I may be wrong, but having a schema based variable declaration seems to be a better solution because it would allow scripts to know that that their build script that uses variables from "Schema A" to use tasks  from another script that use the same "Schema A"

**Comment by nightroman** [comment]
@gaelcolas @RamblingCookieMonster 

A simple and robust but not so obvious technique of exporting and importing module tasks is probably found. See *EXAMPLE 2. Import from a module with tasks* here:
https://github.com/nightroman/Invoke-Build/tree/master/Tasks/Import

**Comment by gaelcolas** [comment]
Not a bad idea and it's lightweight. Will give it a go when I get a chance...

Thanks!

**Comment by gaelcolas** [comment]
ah! I'm confronted again with that good ol' chicken and egg problem....

I like having a single point of entry: Assuming you have `Invoke-Build.ps1` in your repo, you only have to call the `.build.ps1` and then the Build script should find out what it needs by himself (I use a PSDepend task for this called ResolveDependencies).

The problem with resolving those dependencies within a task, is that all the tasks added to, say, the default task `.` needs to be pre-defined...
That means I can't have:
```PowerShell
Param()
if ((Get-PSCallStack)[1].InvocationInfo.MyCommand.Name -ne 'Invoke-Build.ps1') {
    Invoke-Build
    return
}
task . Clean, ResolveDependencies, LoadDependenciesTasks, SomeTaskFromOneDependency, SomeTaskFromTheOtherDependency
task Clean {<# ... #>}
task ResolveDependencies {<# ... #>}
task LoadDependenciesTasks {<# ... #>}
```
Because SomeTaskFromOneDependency and SomeTaskFromTheOtherDependency must be defined *before* running ResolveDependencies...

Any idea how to work around this?

Worst case scenario, I *think* I can start by bootstrapping PSDepend in my `,build.ps1`, resolving the dependencies, before importing the tasks and returning the control to Invoke-Build.

**Comment by gaelcolas** [comment]
Actually, I'm relatively happy with my [bootstrap process in .build.ps1](https://github.com/gaelcolas/SharedDScConfig/blob/master/.build.ps1)

In short I do (or will do), my `.build.ps1` looks like this:
```PowerShell
Param()
Process {
    if ((Get-PSCallStack)[1].InvocationInfo.MyCommand.Name -ne 'Invoke-Build.ps1') {
        Write-Verbose "Returning control to Invoke-Build"
        Invoke-Build
        return
    }
    Resolve-Dependency
    Load-TaskFromLoadedModule
    
}
Begin {
    function Resolve-Dependency { <#...#> }
    function Load-TaskFromLoadedModule {
         (get-modules).exportedAliases.where{key -match '\.tasks$'}.foreach{. $_}
    }

}
```
(I know I should update the invoke-build return with @nightroman updated example)

**Comment by nightroman** [comment]
I think this approach should work.

A side note (up to you): if you prefer to use blocks `Begin` and `Process`, use `End` instead of `Process`. `Process` is designed for processing piped data which we do not have.
